### PR TITLE
feat(metadata-store): #929 Add metadata query to search cache keys

### PR DESCRIPTION
**Pull request by ncloudioj**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/943)

### CODE REVIEW

File: lib/MetadataStore.js [hunk 1]
```diff
@@ -371,6 +371,61 @@ MetadataStore.prototype = {
...
+   *
+   * @param {Array} cacheKeys an cache key array
+   *
+   * Returns a promise with the array of retrieved metadata records
+   */
+  asyncGetMetadataByCacheKey: Task.async(function*(cacheKeys) {
+    const quoted = cacheKeys.map(key => {return `'${key}'`;}).join(",");
+
+    let metaObjects = yield this.asyncExecuteQuery(
+      `SELECT * FROM page_metadata WHERE cache_key IN (${quoted})`,
```
Comment (oyiptong) [inline lib/MetadataStore.js]: Just curious, did you try comparing the performance of one big query vs many small queries?

File: test/test-MetadataStore.js [hunk 2]
```diff
@@ -158,6 +158,22 @@ exports.test_async_insert_all = function*(assert) {
   assert.deepEqual(items[7], [3, 5]);
 };
 
+exports.test_async_get_by_cache_key = function*(assert) {
```
Comment (oyiptong) [inline test/test-MetadataStore.js]: nit: I'd add a couple more tests, if only for regression testing and ensuring the code path has been run:
1. testing for a key that doesn't exist
2. testing for an image type that doesn't exist

File: lib/MetadataStore.js (hunk 1 above)
Comment (ncloudioj) [inline lib/MetadataStore.js]: I haven't done any query benchmark yet. But I did try the one gigantic query, the problem was that you had to sort out the flattened many-to-many result afterward. For instance,

```
metadata1, image1
metadata1, image2
metadata1, favicon1
metadata1, favicon2
...
metadataN, image1
...
metadataN, imageN
```

I think I just chose the simplicity other than performance for now.

File: test/test-MetadataStore.js (hunk 2 above)
Comment (ncloudioj) [inline test/test-MetadataStore.js]: Good idea, will do.

**Pull request merged**